pyo3-derive-backend/src/pyproto.rs

```diff
@@ -91,20 +97,79 @@ fn impl_proto_impl(
             }
         }
     }
+    let inventory_submission = inventory_submission(py_methods, ty);
+    let slot_initialization = slot_initialization(method_names, ty, proto)?;
+    Ok(quote! {
+        #trait_impls
+        #inventory_submission
+        #slot_initialization
+    })
+}
 
+fn inventory_submission(py_methods: Vec<TokenStream>, ty: &syn::Type) -> TokenStream {
     if py_methods.is_empty() {
-        return Ok(quote! { #trait_impls });
+        return quote! {};
     }
-    let inventory_submission = quote! {
+    quote! {
         pyo3::inventory::submit! {
             #![crate = pyo3] {
                 type Inventory = <#ty as pyo3::class::methods::HasMethodsInventory>::Methods;
                 <Inventory as pyo3::class::methods::PyMethodsInventory>::new(&[#(#py_methods),*])
             }
         }
-    };
+    }
+}
+
+fn slot_initialization(
+    method_names: HashSet<String>,
+    ty: &syn::Type,
+    proto: &defs::Proto,
+) -> syn::Result<TokenStream> {
+    let mut initializers: Vec<TokenStream> = vec![];
+    // Some setters cannot coexist.
+    // E.g., if we have `__add__`, we need to skip `set_radd`.
+    let mut skipped_setters = HashSet::new();
```

Comment:
Do you think that we still need `skipped_setters` once we implement a solution for #844?